### DISCUSSION TITLE
Low level pointers VS  c# level objects

### DISCUSSION THREAD

**Discussion by courgeon** [dotnet/Silk.NET#463] [Ideas] [answered]
Hello, First thanks for this awesome work. I've been using several outdates bindings and Silk.NET seems like a great way to unify and keep by bingins up to date.

I'm however running into some difficulties when attempting to migrate my 3D engine : 
Some of GL/GLFW calls are returning low level pointers such as : 
- byte* instead of string (GL.GetString)
- nint instead of string[] (GLFW.DropCallback) (this one was tricky, I had to migrate the entire project to .NET 5.0...)

I feel it's a bit complicated to manipulate these low level pointer from C#. 
Any chances we could get some high level wrapper for these ? Or are there some Utils function somewhere ?

Thanks

**Comment by Perksey** [comment, marked as answer]
Hi, thanks for your feedback!

Bringing our bindings closer to C# and less raw is something we want to work on, however we want to capture the best of both worlds as 1:1 bindings is one of our features.

For `GL.GetString` in particular, there is `GL.GetStringS` which returns a string but this isn't obvious, and in the future I think I'm going to look at adapting that to just overload to have an `out` parameter on the original function. The root problem is we can't overload by return type alone, so I think using `out` parameters will solve that. After that we'll be much more capable of introducing overloads in this fashion.

In GLFW's case, however, these bindings are handwritten and will need more work to fix them.

**Reply by courgeon** [reply]
I missed it, but it's perfectly fine with GetStringS to me. :)
